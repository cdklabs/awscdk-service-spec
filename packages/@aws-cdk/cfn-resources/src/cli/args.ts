export function parseArgv(
  argv: string[],
  namedArgs: string[],
): {
  args: Record<string, string>;
  options: Record<string, string | boolean>;
} {
  return {
    args: Object.fromEntries(
      argv
        .filter((a) => !a.startsWith('--'))
        .map(function (arg, idx) {
          return [namedArgs[idx] ?? idx, arg];
        }),
    ),
    options: Object.fromEntries(
      argv.filter((a) => a.startsWith('--')).map((a) => [a.split('=')[0].substring(2), a.split('=', 2)[1] ?? true]),
    ),
  };
}
