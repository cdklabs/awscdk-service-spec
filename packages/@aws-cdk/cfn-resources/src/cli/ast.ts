import { PropertyType, Resource } from '@aws-cdk/service-spec';
import { Case, InterfaceSpec, MemberKind, TypeReferenceSpec } from '@cdklabs/typewriter';
import * as jsii from '@jsii/spec';

export function resourcePropsSpec(r: Resource): InterfaceSpec {
  const propsInterface = `Cfn${r.name}Props`;

  return {
    export: true,
    name: propsInterface,
    kind: jsii.TypeKind.Interface,
    properties: Object.entries(r.properties).map(([name, p]) => ({
      kind: MemberKind.Property,
      name: Case.firstCharToLower(name),
      type: propertyTypeToTypeReferenceSpec(p.type),
      immutable: true,
    })),
  };
}

export function propertyTypeToTypeReferenceSpec(type: PropertyType): TypeReferenceSpec {
  switch (type?.type) {
    case 'string':
    case 'number':
    case 'boolean':
      return {
        primitive: type.type as jsii.PrimitiveType,
      };
    case 'array':
      return {
        collection: {
          kind: jsii.CollectionKind.Array,
          elementtype: propertyTypeToTypeReferenceSpec(type.element) as any,
        },
      };
    case 'map':
      return {
        collection: {
          kind: jsii.CollectionKind.Map,
          elementtype: propertyTypeToTypeReferenceSpec(type.element) as any,
        },
      };
    case 'json':
    default:
      return {
        primitive: jsii.PrimitiveType.Any,
      };
  }
}
