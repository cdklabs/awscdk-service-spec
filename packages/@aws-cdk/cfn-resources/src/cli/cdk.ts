import { ExternalModule } from '@cdklabs/typewriter';
import { Statement } from '@cdklabs/typewriter/src/statements';

export class CdkCore extends ExternalModule {
  public objectToCloudFormation(...args: Statement[]) {
    return this.import('cdk')
      .invoke('objectToCloudFormation')
      .with(...args);
  }
  public stringToCloudFormation(...args: Statement[]) {
    return this.import('cdk')
      .invoke('stringToCloudFormation')
      .with(...args);
  }
}
