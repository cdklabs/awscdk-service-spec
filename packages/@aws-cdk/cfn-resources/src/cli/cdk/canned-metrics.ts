import { DimensionSet, Metric, Resource, Service, SpecDatabase } from '@aws-cdk/service-spec';
import { ClassType, expr, InterfaceType, IScope, Method, Module, stmt, Type } from '@cdklabs/typewriter';
import { metricFunctionName, metricsClassNameFromService } from '../naming/conventions';

/**
 * Generate Canned Metrics
 */
export class CannedMetricsModule extends Module {
  public static forService(db: SpecDatabase, service: Service): CannedMetricsModule {
    const cm = new CannedMetricsModule(db, service);

    for (const r of db.follow('serviceHasMetric', service)) {
      cm.addMetricWithDimensions(r.entity);
    }

    return cm;
  }

  public static forResource(db: SpecDatabase, resource: Resource): CannedMetricsModule {
    const service = db.incoming('hasResource', resource).only().entity;
    const cm = new CannedMetricsModule(db, service);

    for (const r of db.follow('resourceHasMetric', resource)) {
      cm.addMetricWithDimensions(r.entity);
    }

    return cm;
  }

  private metrics: MetricsClass;
  private _hasCannedMetrics: boolean = false;

  private constructor(private readonly db: SpecDatabase, service: Service) {
    super(`${service.name}.canned-metrics`);
    this.metrics = new MetricsClass(this, service);
  }

  public get hasCannedMetrics() {
    return this._hasCannedMetrics;
  }

  /**
   * Add metrics for a given dimension set to the module
   */
  public addMetricWithDimensions(metric: Metric) {
    this._hasCannedMetrics = true;
    const dimensions = this.db.follow('usesDimensionSet', metric).map((m) => m.entity);
    this.metrics.addMetricWithDimensions(metric, dimensions);
  }
}

export class MetricsClass extends ClassType {
  private returnType: InterfaceType;
  constructor(scope: IScope, service: Service) {
    super(scope, {
      export: true,
      name: metricsClassNameFromService(service),
    });

    this.returnType = new InterfaceType(this.scope, {
      name: 'MetricWithDims',
      properties: [
        {
          name: 'namespace',
          type: Type.STRING,
          immutable: true,
          optional: true,
        },
        {
          name: 'metricName',
          type: Type.STRING,
          immutable: true,
        },
        {
          name: 'statistic',
          type: Type.STRING,
          immutable: true,
        },
      ],
    });

    const dimensionTypeParam = this.returnType.addTypeParameter({ name: 'D' });

    this.returnType.addProperty({
      name: 'dimensionsMap',
      type: dimensionTypeParam.asType(),
      immutable: true,
    });
  }

  public addMetricWithDimensions(metric: Metric, dimensionSets: DimensionSet[]) {
    const name = metricFunctionName(metric);

    // Add a unique declaration for each dimension set
    for (const set of dimensionSets) {
      const dimensionsType = dimensionSetType(set);
      this.addMetricMethodDeclaration(name, dimensionsType);
    }

    // If we have more than one dimension set, add a generic declaration
    if (dimensionSets.length >= 1) {
      this.addMetricMethodDeclaration(name, Type.ANY);
    }

    // Add the implementation to the final declaration
    this.methods.at(-1)?.addBody(
      stmt.ret(
        expr.object({
          namespace: expr.lit(metric.namespace),
          metricName: expr.lit(metric.name),
          dimensionsMap: expr.ident('dimensions'),
          statistic: expr.lit(metric.statistic),
        }),
      ),
    );
  }

  private addMetricMethodDeclaration(name: string, dimensionsType: Type): Method {
    return this.addMethod({
      name,
      static: true,
      returnType: Type.fromName(this.scope, this.returnType.name, [dimensionsType]),
      parameters: [
        {
          name: 'dimensions',
          type: dimensionsType,
        },
      ],
    });
  }
}

function dimensionSetType(set: DimensionSet): Type {
  return Type.anonymousInterface(set.dimensions.map(({ name }) => ({ name, type: Type.STRING })));
}
