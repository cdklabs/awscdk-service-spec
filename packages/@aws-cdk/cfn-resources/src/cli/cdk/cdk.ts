import { $E, $T, expr, Expression, ExternalModule, Scope, ThingSymbol, Type } from '@cdklabs/typewriter';

export interface ModuleImportLocations {
  /**
   * The import name used import the core module
   * @default 'aws-cdk-lib'
   */
  readonly core?: string;
  /**
   * The import name used to import core helpers module
   * @default 'aws-cdk-lib/core/lib/helpers-internal'
   */
  readonly coreHelpers?: string;
}

export class CdkCore extends ExternalModule {
  public readonly helpers = new CdkInternalHelpers(this);

  public readonly CfnResource = Type.fromName(this, 'CfnResource');
  public readonly IInspectable = Type.fromName(this, 'IInspectable');
  public readonly TreeInspector = Type.fromName(this, 'TreeInspector');
  public readonly Token = $T(Type.fromName(this, 'Token'));
  public readonly ResolutionTypeHint = Type.fromName(this, 'ResolutionTypeHint');
  public readonly CfnTag = Type.fromName(this, 'CfnTag');
  public readonly TagManager = $T(Type.fromName(this, 'TagManager'));
  public readonly TagType = $T(Type.fromName(this, 'TagType'));
  public readonly ITaggable = Type.fromName(this, 'ITaggable');

  public readonly objectToCloudFormation = makeCallableExpr(this, 'objectToCloudFormation');
  public readonly stringToCloudFormation = makeCallableExpr(this, 'stringToCloudFormation');
  public readonly dateToCloudFormation = makeCallableExpr(this, 'dateToCloudFormation');
  public readonly booleanToCloudFormation = makeCallableExpr(this, 'booleanToCloudFormation');
  public readonly numberToCloudFormation = makeCallableExpr(this, 'numberToCloudFormation');
  public readonly cfnTagToCloudFormation = makeCallableExpr(this, 'cfnTagToCloudFormation');
  public readonly canInspect = makeCallableExpr(this, 'canInspect');
  public readonly listMapper = makeCallableExpr(this, 'listMapper');
  public readonly hashMapper = makeCallableExpr(this, 'hashMapper');
  public readonly requireProperty = makeCallableExpr(this, 'requireProperty');

  public readonly ValidationResult = $T(Type.fromName(this, 'ValidationResult'));
  public readonly VALIDATION_SUCCESS = makeCallableExpr(this, 'VALIDATION_SUCCESS');
  public readonly ValidationResults = $T(Type.fromName(this, 'ValidationResults'));

  public readonly propertyValidator = makeCallableExpr(this, 'propertyValidator');
  public readonly requiredValidator = makeCallableExpr(this, 'requiredValidator');
  public readonly listValidator = makeCallableExpr(this, 'listValidator');
  public readonly hashValidator = makeCallableExpr(this, 'listValidator');
  public readonly unionValidator = makeCallableExpr(this, 'listValidator');
  public readonly validateCfnTag = makeCallableExpr(this, 'validateCfnTag');
  public readonly validateObject = makeCallableExpr(this, 'validateObject');
  public readonly validateDate = makeCallableExpr(this, 'validateDate');
  public readonly validateBoolean = makeCallableExpr(this, 'validateBoolean');
  public readonly validateNumber = makeCallableExpr(this, 'validateNumber');
  public readonly validateString = makeCallableExpr(this, 'validateString');

  constructor(fqn: string) {
    super(fqn);
  }

  public tokenAsString(arg: Expression) {
    return this.Token.asString(arg);
  }

  public tokenAsNumber(arg: Expression) {
    return this.Token.asNumber(arg);
  }

  public tokenAsList(arg: Expression) {
    return this.Token.asList(arg);
  }
}

export class CdkInternalHelpers extends ExternalModule {
  public readonly FromCloudFormationOptions = Type.fromName(this, 'FromCloudFormationOptions');
  public readonly FromCloudFormationResult = $T(Type.fromName(this, 'FromCloudFormationResult'));
  public readonly FromCloudFormation = $T(Type.fromName(this, 'FromCloudFormation'));
  public readonly FromCloudFormationPropertyObject = Type.fromName(this, 'FromCloudFormationPropertyObject');

  constructor(parent: CdkCore) {
    super(`${parent.fqn}/core/lib/helpers-internal`);
  }
}

export class Constructs extends ExternalModule {
  public readonly Construct = Type.fromName(this, 'Construct');

  constructor() {
    super('constructs');
  }
}

export const CDK_CORE = new CdkCore('aws-cdk-lib');
export const CONSTRUCTS = new Constructs();

function makeCallableExpr(scope: Scope, name: string) {
  return $E(expr.sym(new ThingSymbol(name, scope)));
}
