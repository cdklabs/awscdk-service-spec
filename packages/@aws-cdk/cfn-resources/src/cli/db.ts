import { emptyDatabase } from '@aws-cdk/service-spec';
import fs from 'fs-extra';

const pathToDb = require.resolve('@aws-cdk/service-spec-build/db.json');

export async function loadDatabase() {
  const spec = fs.readJson(pathToDb);
  const db = emptyDatabase();
  db.load(await spec);
  return db;
}
