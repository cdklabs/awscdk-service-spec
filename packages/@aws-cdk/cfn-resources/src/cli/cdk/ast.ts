import { DatabaseSchema, Resource, Service } from '@aws-cdk/service-spec';
import { Database } from '@cdklabs/tskb';
import { StructType, Module } from '@cdklabs/typewriter';
import { Stability } from '@jsii/spec';
import { AugmentationsModule } from './augmentation-generator';
import { CannedMetricsModule } from './canned-metrics';
import { CDK_CORE, CONSTRUCTS, ModuleImportLocations } from './cdk';
import { ResourceClass } from './resource-class';
import { TypeConverter } from './type-converter';
import { classNameFromResource, propStructNameFromResource } from '../naming/conventions';
import { cloudFormationDocLink } from '../naming/doclink';
import { PropMapping } from '../prop-mapping';

/**
 * A module containing a single resource
 */
export class ResourceModule extends Module {
  public constructor(public readonly service: string, public readonly resource: string) {
    super(`@aws-cdk/${service}/${resource}-l1`);
  }
}

/**
 * A module containing a service
 */
export class ServiceModule extends Module {
  public constructor(public readonly service: string, public readonly shortName: string) {
    super(`@aws-cdk/${service}`);
  }
}

export interface AstBuilderProps {
  readonly db: Database<DatabaseSchema>;
  /**
   * Override the locations modules are imported from
   */
  readonly importLocations?: ModuleImportLocations;
}

export class AstBuilder<T extends Module> {
  /**
   * Build a module for all resources in a service
   */
  public static forService(service: Service, props: AstBuilderProps): AstBuilder<ServiceModule> {
    const scope = new ServiceModule(service.name, service.shortName);
    const aug = new AugmentationsModule(props.db, service.name, props.importLocations?.cloudwatch);
    const metrics = CannedMetricsModule.forService(props.db, service);

    const ast = new AstBuilder(scope, props, aug, metrics);

    const resources = props.db.follow('hasResource', service);

    for (const link of resources) {
      ast.addResource(link.entity);
    }

    return ast;
  }

  /**
   * Build an module for a single resource
   */
  public static forResource(resource: Resource, props: AstBuilderProps): AstBuilder<ResourceModule> {
    const parts = resource.cloudFormationType.toLowerCase().split('::');
    const scope = new ResourceModule(parts[1], parts[2]);
    const aug = new AugmentationsModule(props.db, parts[1], props.importLocations?.cloudwatch);
    const metrics = CannedMetricsModule.forResource(props.db, resource);

    const ast = new AstBuilder(scope, props, aug, metrics);
    ast.addResource(resource);

    return ast;
  }

  public readonly db: Database<DatabaseSchema>;

  protected constructor(
    public readonly module: T,
    props: AstBuilderProps,
    public readonly augmentations?: AugmentationsModule,
    public readonly cannedMetrics?: CannedMetricsModule,
  ) {
    this.db = props.db;

    CDK_CORE.import(module, 'cdk', { fromLocation: props.importLocations?.core });
    CONSTRUCTS.import(module, 'constructs');
    CDK_CORE.helpers.import(module, 'cfn_parse', { fromLocation: props.importLocations?.coreHelpers });
  }

  public addResource(resource: Resource) {
    const resourceClass = new ResourceClass(this.module, resource);

    const converter = new TypeConverter({ db: this.db, resource, resourceClass });
    const propsType = this.addResourcePropsType(resource, converter);
    resourceClass.buildMembers(propsType);

    this.augmentations?.augmentResource(resource, resourceClass);
  }

  protected addResourcePropsType(r: Resource, converter: TypeConverter) {
    const propsInterface = new StructType(this.module, {
      export: true,
      name: propStructNameFromResource(r),
      docs: {
        summary: `Properties for defining a \`${classNameFromResource(r)}\``,
        stability: Stability.External,
        see: cloudFormationDocLink({
          resourceType: r.cloudFormationType,
        }),
      },
    });
    const mapping = new PropMapping(this.module);
    for (const [name, prop] of Object.entries(r.properties)) {
      converter.addStructProperty(propsInterface, mapping, name, prop);
    }

    converter.makeCfnProducer(propsInterface, mapping);
    converter.makeCfnParser(propsInterface, mapping);
    return propsInterface;
  }
}
