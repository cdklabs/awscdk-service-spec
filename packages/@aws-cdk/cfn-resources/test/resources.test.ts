import { DatabaseSchema } from '@aws-cdk/service-spec';
import { Database } from '@cdklabs/tskb';
import { TypeScriptRenderer } from '@cdklabs/typewriter';
import { AstBuilder } from '../src/cli/cdk/ast';
import { loadDatabase } from '../src/cli/db';

const renderer = new TypeScriptRenderer();
let db: Database<DatabaseSchema>;

beforeAll(async () => {
  db = await loadDatabase();
});

test.each([
  'Alexa::ASK::Skill',
  'AWS::ApiGateway::RestApi',
  'AWS::IAM::Role',
  'AWS::Lambda::Function',
  'AWS::S3::Bucket',
  'AWS::SQS::Queue',
])('%s', (cloudFormationType) => {
  const resource = db.lookup('resource', 'cloudFormationType', 'equals', cloudFormationType)[0];

  const ast = AstBuilder.forResource(resource, { db });

  expect(renderer.render(ast.scope)).toMatchSnapshot();
});
