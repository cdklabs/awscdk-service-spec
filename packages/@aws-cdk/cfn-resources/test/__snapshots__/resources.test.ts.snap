// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWS::ApiGateway::RestApi 1`] = `
"export interface BodyS3Location {
  readonly bucket: string;

  readonly eTag: string;

  readonly version: string;

  readonly key: string;
}


export interface EndpointConfiguration {
  readonly types: Array<string>;

  readonly vpcEndpointIds: Array<string>;
}


export interface Tags {
  readonly key: string;

  readonly value: string;
}


export interface CfnRestApiProps {
  readonly apiKeySourceType: string;

  readonly binaryMediaTypes: Array<string>;

  readonly body: any;

  readonly bodyS3Location: BodyS3Location;

  readonly cloneFrom: string;

  readonly endpointConfiguration: EndpointConfiguration;

  readonly description: string;

  readonly disableExecuteApiEndpoint: boolean;

  readonly failOnWarnings: boolean;

  readonly name: string;

  readonly minimumCompressionSize: number;

  readonly mode: string;

  readonly policy: any;

  readonly parameters: any;

  readonly tags: Array<Tags>;
}
"
`;

exports[`AWS::IAM::Role 1`] = `
"export interface Policies {
  readonly policyDocument: any;

  readonly policyName: string;
}


export interface Tags {
  readonly key: string;

  readonly value: string;
}


export interface CfnRoleProps {
  readonly assumeRolePolicyDocument: any;

  readonly description: string;

  readonly managedPolicyArns: Array<string>;

  readonly maxSessionDuration: number;

  readonly path: string;

  readonly permissionsBoundary: string;

  readonly policies: Array<Policies>;

  readonly roleName: string;

  readonly tags: Array<Tags>;
}
"
`;

exports[`AWS::Lambda::Function 1`] = `
"export interface TracingConfig {
  readonly mode: string;
}


export interface VpcConfig {
  readonly securityGroupIds: Array<string>;

  readonly subnetIds: Array<string>;
}


export interface SnapStart {
  readonly applyOn: string;
}


export interface FileSystemConfigs {
  readonly arn: string;

  readonly localMountPath: string;
}


export interface Tags {
  readonly value: string;

  readonly key: string;
}


export interface ImageConfig {
  readonly workingDirectory: string;

  readonly command: Array<string>;

  readonly entryPoint: Array<string>;
}


export interface DeadLetterConfig {
  readonly targetArn: string;
}


export interface Code {
  readonly s3ObjectVersion: string;

  readonly s3Bucket: string;

  readonly zipFile: string;

  readonly s3Key: string;

  readonly imageUri: string;
}


export interface Environment {
  readonly variables: Map<string, string>;
}


export interface EphemeralStorage {
  readonly size: number;
}


export interface CfnFunctionProps {
  readonly description: string;

  readonly tracingConfig: TracingConfig;

  readonly vpcConfig: VpcConfig;

  readonly reservedConcurrentExecutions: number;

  readonly snapStart: SnapStart;

  readonly fileSystemConfigs: Array<FileSystemConfigs>;

  readonly functionName: string;

  readonly runtime: string;

  readonly kmsKeyArn: string;

  readonly packageType: string;

  readonly codeSigningConfigArn: string;

  readonly layers: Array<string>;

  readonly tags: Array<Tags>;

  readonly imageConfig: ImageConfig;

  readonly memorySize: number;

  readonly deadLetterConfig: DeadLetterConfig;

  readonly timeout: number;

  readonly handler: string;

  readonly code: Code;

  readonly role: string;

  readonly environment: Environment;

  readonly ephemeralStorage: EphemeralStorage;

  readonly architectures: Array<string>;
}
"
`;

exports[`AWS::S3::Bucket 1`] = `
"export interface AccelerateConfiguration {
  readonly accelerationStatus: string;
}


export interface AnalyticsConfigurations {
  readonly tagFilters: Array<TagFilters>;

  readonly storageClassAnalysis: StorageClassAnalysis;

  readonly id: string;

  readonly prefix: string;
}


export interface TagFilters {
  readonly value: string;

  readonly key: string;
}


export interface StorageClassAnalysis {
  readonly dataExport: DataExport;
}


export interface DataExport {
  readonly destination: Destination;

  readonly outputSchemaVersion: string;
}


export interface Destination {
  readonly bucketArn: string;

  readonly bucketAccountId: string;

  readonly format: string;

  readonly prefix: string;
}


export interface BucketEncryption {
  readonly serverSideEncryptionConfiguration: Array<ServerSideEncryptionConfiguration>;
}


export interface ServerSideEncryptionConfiguration {
  readonly bucketKeyEnabled: boolean;

  readonly serverSideEncryptionByDefault: ServerSideEncryptionByDefault;
}


export interface ServerSideEncryptionByDefault {
  readonly kMSMasterKeyID: string;

  readonly sSEAlgorithm: string;
}


export interface CorsConfiguration {
  readonly corsRules: Array<CorsRules>;
}


export interface CorsRules {
  readonly allowedHeaders: Array<string>;

  readonly allowedMethods: Array<string>;

  readonly allowedOrigins: Array<string>;

  readonly exposedHeaders: Array<string>;

  readonly id: string;

  readonly maxAge: number;
}


export interface IntelligentTieringConfigurations {
  readonly id: string;

  readonly prefix: string;

  readonly status: string;

  readonly tagFilters: Array<TagFilters>;

  readonly tierings: Array<Tierings>;
}


export interface Tierings {
  readonly accessTier: string;

  readonly days: number;
}


export interface InventoryConfigurations {
  readonly destination: Destination;

  readonly enabled: boolean;

  readonly id: string;

  readonly includedObjectVersions: string;

  readonly optionalFields: Array<string>;

  readonly prefix: string;

  readonly scheduleFrequency: string;
}


export interface LifecycleConfiguration {
  readonly rules: Array<Rules>;
}


export interface Rules {
  readonly abortIncompleteMultipartUpload: AbortIncompleteMultipartUpload;

  readonly expirationDate: string;

  readonly expirationInDays: number;

  readonly expiredObjectDeleteMarker: boolean;

  readonly id: string;

  readonly noncurrentVersionExpirationInDays: number;

  readonly noncurrentVersionExpiration: NoncurrentVersionExpiration;

  readonly noncurrentVersionTransition: NoncurrentVersionTransition;

  readonly noncurrentVersionTransitions: Array<NoncurrentVersionTransition>;

  readonly prefix: string;

  readonly status: string;

  readonly tagFilters: Array<TagFilters>;

  readonly objectSizeGreaterThan: string;

  readonly objectSizeLessThan: string;

  readonly transition: Transition;

  readonly transitions: Array<Transition>;
}


export interface AbortIncompleteMultipartUpload {
  readonly daysAfterInitiation: number;
}


export interface NoncurrentVersionExpiration {
  readonly noncurrentDays: number;

  readonly newerNoncurrentVersions: number;
}


export interface NoncurrentVersionTransition {
  readonly storageClass: string;

  readonly transitionInDays: number;

  readonly newerNoncurrentVersions: number;
}


export interface Transition {
  readonly storageClass: string;

  readonly transitionDate: string;

  readonly transitionInDays: number;
}


export interface LoggingConfiguration {
  readonly destinationBucketName: string;

  readonly logFilePrefix: string;
}


export interface MetricsConfigurations {
  readonly accessPointArn: string;

  readonly id: string;

  readonly prefix: string;

  readonly tagFilters: Array<TagFilters>;
}


export interface NotificationConfiguration {
  readonly eventBridgeConfiguration: EventBridgeConfiguration;

  readonly lambdaConfigurations: Array<LambdaConfigurations>;

  readonly queueConfigurations: Array<QueueConfigurations>;

  readonly topicConfigurations: Array<TopicConfigurations>;
}


export interface EventBridgeConfiguration {
  readonly eventBridgeEnabled: boolean;
}


export interface LambdaConfigurations {
  readonly event: string;

  readonly filter: Filter;

  readonly function: string;
}


export interface Filter {
  readonly s3Key: S3Key;
}


export interface S3Key {
  readonly rules: Array<Rules>;
}


export interface QueueConfigurations {
  readonly event: string;

  readonly filter: Filter;

  readonly queue: string;
}


export interface TopicConfigurations {
  readonly event: string;

  readonly filter: Filter;

  readonly topic: string;
}


export interface ObjectLockConfiguration {
  readonly objectLockEnabled: string;

  readonly rule: Rule;
}


export interface Rule {
  readonly defaultRetention: DefaultRetention;
}


export interface DefaultRetention {
  readonly years: number;

  readonly days: number;

  readonly mode: string;
}


export interface OwnershipControls {
  readonly rules: Array<Rules>;
}


export interface PublicAccessBlockConfiguration {
  readonly blockPublicAcls: boolean;

  readonly blockPublicPolicy: boolean;

  readonly ignorePublicAcls: boolean;

  readonly restrictPublicBuckets: boolean;
}


export interface ReplicationConfiguration {
  readonly role: string;

  readonly rules: Array<Rules>;
}


export interface Tags {
  readonly key: string;

  readonly value: string;
}


export interface VersioningConfiguration {
  readonly status: string;
}


export interface WebsiteConfiguration {
  readonly errorDocument: string;

  readonly indexDocument: string;

  readonly routingRules: Array<RoutingRules>;

  readonly redirectAllRequestsTo: RedirectAllRequestsTo;
}


export interface RoutingRules {
  readonly redirectRule: RedirectRule;

  readonly routingRuleCondition: RoutingRuleCondition;
}


export interface RedirectRule {
  readonly hostName: string;

  readonly httpRedirectCode: string;

  readonly protocol: string;

  readonly replaceKeyPrefixWith: string;

  readonly replaceKeyWith: string;
}


export interface RoutingRuleCondition {
  readonly keyPrefixEquals: string;

  readonly httpErrorCodeReturnedEquals: string;
}


export interface RedirectAllRequestsTo {
  readonly hostName: string;

  readonly protocol: string;
}


export interface CfnBucketProps {
  readonly accelerateConfiguration: AccelerateConfiguration;

  readonly accessControl: string;

  readonly analyticsConfigurations: Array<AnalyticsConfigurations>;

  readonly bucketEncryption: BucketEncryption;

  readonly bucketName: string;

  readonly corsConfiguration: CorsConfiguration;

  readonly intelligentTieringConfigurations: Array<IntelligentTieringConfigurations>;

  readonly inventoryConfigurations: Array<InventoryConfigurations>;

  readonly lifecycleConfiguration: LifecycleConfiguration;

  readonly loggingConfiguration: LoggingConfiguration;

  readonly metricsConfigurations: Array<MetricsConfigurations>;

  readonly notificationConfiguration: NotificationConfiguration;

  readonly objectLockConfiguration: ObjectLockConfiguration;

  readonly objectLockEnabled: boolean;

  readonly ownershipControls: OwnershipControls;

  readonly publicAccessBlockConfiguration: PublicAccessBlockConfiguration;

  readonly replicationConfiguration: ReplicationConfiguration;

  readonly tags: Array<Tags>;

  readonly versioningConfiguration: VersioningConfiguration;

  readonly websiteConfiguration: WebsiteConfiguration;
}
"
`;

exports[`AWS::SQS::Queue 1`] = `
"export interface Tags {
  readonly key: string;

  readonly value: string;
}


export interface CfnQueueProps {
  readonly contentBasedDeduplication: boolean;

  readonly deduplicationScope: string;

  readonly delaySeconds: number;

  readonly fifoQueue: boolean;

  readonly fifoThroughputLimit: string;

  readonly kmsDataKeyReusePeriodSeconds: number;

  readonly kmsMasterKeyId: string;

  readonly sqsManagedSseEnabled: boolean;

  readonly maximumMessageSize: number;

  readonly messageRetentionPeriod: number;

  readonly queueName: string;

  readonly receiveMessageWaitTimeSeconds: number;

  readonly redriveAllowPolicy: any;

  readonly redrivePolicy: any;

  readonly tags: Array<Tags>;

  readonly visibilityTimeout: number;
}
"
`;

exports[`Alexa::ASK::Skill 1`] = `
"export interface AuthenticationConfiguration {
  readonly clientId: string;

  readonly refreshToken: string;

  readonly clientSecret: string;
}


export interface SkillPackage {
  readonly s3BucketRole: string;

  readonly overrides: Overrides;

  readonly s3ObjectVersion: string;

  readonly s3Bucket: string;

  readonly s3Key: string;
}


export interface Overrides {
  readonly manifest: any;
}


export interface CfnSkillProps {
  readonly authenticationConfiguration: AuthenticationConfiguration;

  readonly vendorId: string;

  readonly skillPackage: SkillPackage;
}
"
`;
